docs/proposals/015-bindless-t.md

```diff
@@ -0,0 +1,160 @@
+    let t = Bindless<Texture2D>(idx);
+
+    // Comparison.
+    ASSERT(t == texture);
+
+    // OK, `t` is first implicitly dereferenced to producee `Texture2D`, and
+    // then `Texture2D::Sample` is called.
+    // The `sampler` argument is implicitly converted from `Bindless<SamplerState>`
+    // to `SamplerState`.
+    t.Sample(sampler, float2(0,0));
+}
+```
+
+A `Bindless<T>` type is a concrete type whose size is always 8 bytes and is internally a 64-bit integer.
+This means that you can use a `Bindless<T>` type in any context where an ordinary data type, e.g. `int` type
+is allowed, such as in buffer elements.
+
+On targets where resource handles are already concrete and sized types, `Bindless<T>` simply translates to just `T`.
+
+### Obtaining Actual Resource Handle from `Bindless<T>`
+
+Depending on the target platform and the design choices of the user's application, the way to obtain the actual
+resource handle from a `Bindless<T>` integer handle can vary. Slang does not dictate how this conversion is done,
+and instead, this is left to the user via Slang's link-time specialization ability.
+
+Slang defines the following core module declarations:
+
+```slang
+extern T getResourceFromBindlessHandle(uint64_t handle) where T : IOpaqueHandle
+{
+    // Default Implementation
+    // ...
+}
+```
+
+The `getResourceFromBindlessHandle` is used to convert from a bindless handle to actual opaque resource handle.
+If this function is not provided by the user, the default implementation defined in the core module will be used.
+
+By default, the core module implementation of `getResourceFromBindlessHandle` should use the `ResourceDescriptorHeap` and
+`SamplerDescriptorHeap` builtin object when generating HLSL code. When generating code on other targets, `getResourceFromBindlessHandle`
+will fetch the resource handle from a system defined global array of the corresponding resource type.
+
+If/when SPIRV is extended to expose similar capabilities as D3D's `ResourceDescriptorHeap` feature, we should change the default implementation
+to use that instead. Until we know the default implementation of `getResourceFromBindlessHandle` is stable, we should advise users
+to provide their own implementation of `getResourceFromBindlessHandle` to prevent breakages.
+
+If the user application requires a different bindless implementation, this default behavior can be overrided by defining
+`getResourceFromBindlessHandle` in the user code. Below is a possible user-space implementation of `getResourceFromBindlessHandle`
+for Vulkan:
+
+```slang
+
+// All texture and buffer handles are defined in descriptor set 100.
+[vk::binding(0, 100)]
+__DynamicResource<__DynamicResourceKind.General> resourceHandles[];
+
+// All sampler handles are defined in descriptor set 101.
+[vk::binding(0, 101)]
+__DynamicResource<__DynamicResourceKind.Sampler> samplerHandles[];
+
+export getResourceFromBindlessHandle<T>(uint64_t handle) where T : IOpaqueHandle
```

Comment:
I'm behind on reading up on the link-time specialization feature, but is there any note that needs to be made on what happens if multiple user modules try to implement this method?

I'm not entirely comfortable with this kind of thing as a mechanism, but I also have to concede that I can't think of anything better that we could implement in a reasonable timeframe.